File: problems/12_advanced_graphs/0778_swim_in_rising_water/solution.py
"""
Problem: 0778. Swim in Rising Water
Difficulty: Hard
URL: https://leetcode.com/problems/swim-in-rising-water/

解法：Dijkstra（min-max）。
"""
from __future__ import annotations
from typing import List
import heapq


class Solution:
    def swimInWater(self, grid: List[List[int]]) -> int:
        n = len(grid)
        visited = [[False] * n for _ in range(n)]
        h = [(grid[0][0], 0, 0)]
        while h:
            t, i, j = heapq.heappop(h)
            if i == n - 1 and j == n - 1:
                return t
            if visited[i][j]:
                continue
            visited[i][j] = True
            for di, dj in ((1, 0), (-1, 0), (0, 1), (0, -1)):
                ni, nj = i + di, j + dj
                if 0 <= ni < n and 0 <= nj < n and not visited[ni][nj]:
                    heapq.heappush(h, (max(t, grid[ni][nj]), ni, nj))
        return -1


if __name__ == "__main__":
    sol = Solution()
    assert sol.swimInWater([[0, 2], [1, 3]]) == 3
    assert sol.swimInWater([
        [0, 1, 2, 3, 4],
        [24, 23, 22, 21, 5],
        [12, 13, 14, 15, 16],
        [11, 17, 18, 19, 20],
        [10, 9, 8, 7, 6],
    ]) == 16
    assert sol.swimInWater([[0]]) == 0
    assert sol.swimInWater([[0, 1], [2, 3]]) == 3
    assert sol.swimInWater([[3, 2], [0, 1]]) == 3
    assert sol.swimInWater([[10, 12], [20, 15]]) == 15
    print("All tests passed!")

File: problems/12_advanced_graphs/0778_swim_in_rising_water/solution.swift
// Problem: 0778. Swim in Rising Water
// Difficulty: Hard

import Foundation

struct MinHeap778 {
    private var a: [(Int, Int, Int)] = []
    var isEmpty: Bool { a.isEmpty }
    mutating func push(_ v: (Int, Int, Int)) { a.append(v); up(a.count - 1) }
    mutating func pop() -> (Int, Int, Int) {
        let t = a[0]; let last = a.removeLast()
        if !a.isEmpty { a[0] = last; down(0) }
        return t
    }
    private mutating func up(_ i: Int) {
        // TODO: implement
    }
    private mutating func down(_ i: Int) {
        // TODO: implement
    }
}

class Solution {
    func swimInWater(_ grid: [[Int]]) -> Int {
        // TODO: implement
        return 0
    }
}

func checkInt(_ a: Int, _ e: Int, _ msg: String) {
    if a != e { fatalError("FAIL \(msg): \(a) vs \(e)") }
}

// MARK: - Tests
// let s = Solution()

// Test 1
// let result = s.solve(...)

print("All tests passed!")
